faker/providers/barcode/__init__.py

```diff
@@ -23,8 +58,157 @@ def ean(self, length=13):
 
         return ''.join(str(x) for x in code)
 
+    def _convert_upc_a2e(self, upc_a):
+        """
+        Converts a 12-digit UPC-A barcode to its 8-digit UPC-E equivalent
+
+        Not all UPC-A barcodes can be converted.
+
+        :param upc_a: UPC-A barcode to convert
+        :return: UPC-E equivalent barcode
+        """
+        if not isinstance(upc_a, string_types):
+            raise TypeError('`upc_a` is not a string')
+        m1 = self.upc_ae_pattern1.match(upc_a)
+        m2 = self.upc_ae_pattern2.match(upc_a)
+        m3 = self.upc_ae_pattern3.match(upc_a)
+        if not any([m1, m2, m3]):
+            raise ValueError('`upc_a` has an invalid value')
+        upc_e_template = '{number_system_digit}{mfr_code}{product_code}{extra}{check_digit}'
+        if m1:
+            upc_e = upc_e_template.format(**m1.groupdict())
+        elif m2:
+            groupdict = m2.groupdict()
+            groupdict['extra'] = str(len(groupdict.get('mfr_code')))
+            upc_e = upc_e_template.format(**groupdict)
+        else:
+            groupdict = m3.groupdict()
+            groupdict['product_code'] = ''
+            upc_e = upc_e_template.format(**groupdict)
+        return upc_e
+
+    def _upc_ae(self, *, base=None, number_system_digit=None):
+        """
+        Creates a 12-digit UPC-A barcode that can be converted to UPC-E
+
+        Notes on this method:
+        - Please view notes on `upc_a()` and `upc_e()` first.
+
+        :param base: A 6-digit string
+        :param number_system_digit: 0, 1, '0', '1'
+        :return: 12-digit UPC-A barcode that can be converted to UPC-E
+        """
+        if isinstance(base, string_types) and self.upc_e_base_pattern.match(base):
+            base = [int(x) for x in base]
+        else:
+            base = [self.random_int(0, 9) for _ in range(6)]
+        if number_system_digit in ['0', '1', 0, 1]:
+            number_system_digit = int(number_system_digit)
+        else:
+            number_system_digit = self.random_int(0, 1)
+
+        if base[-1] <= 2:
+            code = base[:2] + base[-1:] + [0] * 4 + base[2:-1]
+        elif base[-1] <= 4:
+            code = base[:base[-1]] + [0] * 5 + base[base[-1]:-1]
+        else:
+            code = base[:5] + [0] * 4 + base[-1:]
+
+        code.insert(0, number_system_digit)
+        weights = [3, 1, 3, 1, 3, 1, 3, 1, 3, 1, 3]
+        weighted_sum = sum(x * y for x, y in zip(code, weights))
+        check_digit = (10 - weighted_sum % 10) % 10
+        code.append(check_digit)
+        return ''.join(str(x) for x in code)
+
+    def ean(self, length=13):
+        return self._ean(length)
+
     def ean8(self):
-        return self.ean(8)
+        return self._ean(8)
+
+    def ean13(self, *, no_leading_zero=False, force_leading_zero=False):
```

Comment:
I'm not sure why we should use two arguments. Couldn't we just have `leading_zero`?

I think this will behavior would be more natural:

* `leading_zero=None`: the leading digit could be anything
* `leading_zero=True`: the leading digit will be `0`
* `leading_zero=False`: the leading digit will be 1-9

Reply:
I actually thought about that, and for some reason, it felt incredibly unnatural to me at the time. Now that you point it out, I do not know what in the world "past me" was thinking. I will fix this.

Reply:
Don't worry, we all have those moments 😃  That's what code reviews are for!